### DISCUSSION TITLE
HC-SR04 Obstacle Avoiding

### DISCUSSION THREAD

**Discussion by WhoIsMrSentry** [makerforgetech/modular-biped#71] [Ideas]
Hello Community, an idea suddenly came to my mind. Why don't we add an HC-SR04 to the robot? It is a fact that we need to activate the walking function so that it can automatically move away when it gets too close to a place or a person.

My opinion is that since the piece (hip) that connects the two feet looks directly at the ground, if there was a HC-SR04 there, it would see the obstacles in front of it while walking, with the HC-SR04 in the hip and the HC-SR04 in the part where the neck and body are.

![Front](https://github.com/makerforgetech/modular-biped/assets/134494889/01652980-b003-4f97-b3b4-08017f1f292d)

![Hip](https://github.com/makerforgetech/modular-biped/assets/134494889/aee37f3e-01ca-459c-9c26-510073542a9f)

![Right](https://github.com/makerforgetech/modular-biped/assets/134494889/d4bab6ab-16fb-4472-8bad-94198d989778)
Maybe it would be a good idea to add left and right

Sounds like a useful idea once the walking function is activated.
I'm curious about your thoughts and advice.

**Comment by danic85** [comment]
Definitely a possibility, you can use a multiplexer to reduce the number of pins used for the collection of sensors.

**Reply by WhoIsMrSentry** [reply]
I have heard a lot about multiplexer modules, but I have never had the opportunity to use them. If you have knowledge, can you elaborate on the subject a little?

**Reply by danic85** [reply]
It's a pretty simple concept that uses one or more pins as selectors so that you can use a single input pin to read from multiple sensors.
The number of selector pins increases with the number of 'channels', but it's still less than connecting those channels to input pins on your microcontroller.

For example, you can connect 16 devices with 4 selector pins and a single input pin. 

https://www.electronics-tutorials.ws/combination/comb_2.html

https://mou.sr/3tJRMUA

**Comment by ScottW1** [comment]
They do make i2c versions of that module. 
Scott Winslow

**Reply by danic85** [reply]
I understood with i2c that you're limited to the number of addresses you can set the device to... For example on [this](https://www.adafruit.com/product/4742) i2c ultrasonic sensor it uses the address 0x57 which cannot be changed, so installing more than one on the same i2c bus would cause an address conflict (which would need a multiplexer to resolve anyway).

Ref: https://www.quora.com/How-do-you-connect-multiple-I2C-devices-with-the-same-address

**Comment by ScottW1** [comment]
According to this chat board, you can change the address in software. I have not tried it though to confirm.How to change I2C Address from Ultrasonic Sensor SU04discuss.ardupilot.orgAlthough to be honest, I’m not totally sure why you would need more than one. A single unit mounted on the main body between the legs should give a decent “view angle” for obstructions. You might want to look at the tmf8820 laser “3d sensor” it’s so much smaller than the ultrasonic devices and doesn’t make the annoying humming noise when active. It looks like although they have a locked i2c address normally, they have an enable pin that can switch between multiple scanners. Of course that may get you back to a multiplexer again. Ha ha! But maybe just a single enable pin instead of the trigger and receive pins. Personally I’m working on integrating an OAK-D lite camera system from Luxonis for facial recognition, object detection and distance to object calculation. It allows me to offload the heavy processing to the camera’s built in processors, and has been working fairly smoothly with the pi4. Scott On Jan 8, 2024, at 11:22 AM, Dan Nicholson ***@***.***> wrote:﻿
It's a pretty simple concept that uses one or more pins as selectors so that you can use a single input pin to read from multiple sensors.
The number of selector pins increases with the number of 'channels', but it's still less than connecting those channels to input pins on your microcontroller.
For example, you can connect 16 devices with 4 selector pins and a single input pin.
https://www.electronics-tutorials.ws/combination/comb_2.html
https://www.mouser.co.uk/ProductDetail/SparkFun/BOB-09056
